packages/headless/src/features/search-parameters/search-parameter-serializer.ts

```diff
@@ -224,8 +232,23 @@ function cast<K extends keyof SearchParameters>(
   }
 
   if (key === 'f' || key === 'cf' || key === 'nf' || key === 'df') {
-    return [key, JSON.parse(value)];
+    return [key, castUnknownFacetObject(value)];
   }
 
-  return pair;
+  return [key, decodeURIComponent(value)];
+}
+
+function castUnknownFacetObject(value: string) {
+  const jsonParsed = JSON.parse(value) as unknownFacetObject;
+  const ret = {} as unknownFacetObject;
+  Object.entries(jsonParsed).forEach((entry) => {
+    ret[entry[0]] = entry[1].map((e) => {
```

Comment:
I think the code would be easier to follow if the values were destructured into variables:

```
const [facetId, values] = entry;
ret[facetId] = values.map(v => isString(v) ? decodeURIComponent(e) : v)
```